cpp/src/arrow/io/file_test.cc

```diff
@@ -635,6 +635,41 @@ TEST_F(TestMemoryMappedFile, ZeroSizeFlie) {
   ASSERT_EQ(0, size);
 }
 
+TEST_F(TestMemoryMappedFile, MapPartFile) {
+  const int64_t buffer_size = 1024;
+  const int64_t unalign_offset = 1024;
+  const int64_t offset = 65536;
+  std::vector<uint8_t> buffer(buffer_size);
+
+  random_bytes(1024, 0, buffer.data());
+
+  const int reps = 128;
+
+  std::string path = "io-memory-map-offset";
+  std::shared_ptr<MemoryMappedFile> result;
+
+  // file size = 128k
+  CreateFile(path, reps * buffer_size);
+
+  // map failed with unaligned offset
+  ASSERT_RAISES(IOError, MemoryMappedFile::Open(path, FileMode::READWRITE, 4096,
+                                                unalign_offset, &result));
+
+  // map file region <64k-68k>
+  ASSERT_OK(MemoryMappedFile::Open(path, FileMode::READWRITE, 4096, offset, &result));
+
+  std::shared_ptr<Buffer> out_buffer;
+  ASSERT_OK(result->Write(buffer.data(), buffer_size));
+  ASSERT_OK(result->ReadAt(0, buffer_size, &out_buffer));
+  ASSERT_EQ(0, memcmp(out_buffer->data(), buffer.data(), buffer_size));
+
+  // Resize is not supported
+  ASSERT_RAISES(IOError, result->Resize(4096));
+
+  // Write beyond memory mapped length
+  ASSERT_RAISES(Invalid, result->WriteAt(4096, buffer.data(), buffer_size));
```

Comment:
Should you check `GetSize()` and `Seek()` as well?

Reply:
Tests on GetSize(), Seek(), Tell() added.